src/main/java/com/iota/iri/storage/rocksDB/RocksDBPersistenceProvider.java

```diff
@@ -83,438 +88,373 @@ public boolean isAvailable() {
-            object.readMetadata(db.get(referenceHandle, index == null? new byte[0]: index.bytes()));
+        object.read(db.get(classTreeMap.get(model), index == null ? new byte[0] : index.bytes()));
+
+        ColumnFamilyHandle referenceHandle = metadataReference.get(model);
+        if (referenceHandle != null) {
+            object.readMetadata(db.get(referenceHandle, index == null ? new byte[0] : index.bytes()));
         }
         return object;
     }
 
-
     @Override
-    public boolean mayExist(Class<?> model, Indexable index) throws Exception {
-        ColumnFamilyHandle handle = classTreeMap.get().get(model);
+    public boolean mayExist(Class<?> model, Indexable index) {
+        ColumnFamilyHandle handle = classTreeMap.get(model);
         return db.keyMayExist(handle, index.bytes(), new StringBuilder());
     }
 
     @Override
     public long count(Class<?> model) throws Exception {
         return getCountEstimate(model);
-        //return counts.get(model);
     }
 
     private long getCountEstimate(Class<?> model) throws RocksDBException {
-        ColumnFamilyHandle handle = classTreeMap.get().get(model);
+        ColumnFamilyHandle handle = classTreeMap.get(model);
         return db.getLongProperty(handle, "rocksdb.estimate-num-keys");
     }
 
     @Override
     public Set<Indexable> keysStartingWith(Class<?> modelClass, byte[] value) {
-        RocksIterator iterator;
-        ColumnFamilyHandle handle = classTreeMap.get().get(modelClass);
-        Set<Indexable> keys = new HashSet<>();
-        if(handle != null) {
-            iterator = db.newIterator(handle);
-            try {
+        Objects.requireNonNull(value, "value byte[] cannot be null");
+        ColumnFamilyHandle handle = classTreeMap.get(modelClass);
+        Set<Indexable> keys = null;
+        if (handle != null) {
+            try (RocksIterator iterator = db.newIterator(handle)) {
                 iterator.seek(new Hash(value, 0, value.length).bytes());
-                for(;
-                    iterator.isValid() && Arrays.equals(Arrays.copyOf(iterator.key(), value.length), value);
-                    iterator.next()) {
-                    keys.add(new Hash(iterator.key()));
+
+                byte[] found;
+                while (iterator.isValid() && compareStartsWithKey(value, found = iterator.key())) {
+                    keys = keys == null ? new HashSet<>() : keys;
+                    keys.add(new Hash(found));
+                    iterator.next();
                 }
-            } finally {
-                iterator.close();
             }
         }
-        return keys;
+        return keys == null ? EMPTY_SET : Collections.unmodifiableSet(keys);
+    }
+
+    /**
+     * @param key   What we are looking for.
+     * @param value The bytes we are searching in.
+     * @return true If the value byte[] starts with the key byte[].
+     */
+    private boolean compareStartsWithKey(byte[] key, byte[] value) {
```

Comment:
When I see a method starts with `compare` it confuses me (I expect it to return an int).
Also it is a private method that is called with the parameters `(value, iterator.key())`, however it recieves `(key, value)`,
I find it also confusing.

Since code is written once but read many times please change to `keyStartsWithValue(byte[] value, byte[] key)`.

Also, this method can (and thus should)  be static.